storage/innobase/fil/fil0crypt.cc

```diff
@@ -0,0 +1,3067 @@
+	* crypt data in page 0 */
+
+	crypt_data = fil_space_create_crypt_data(FIL_ENCRYPTION_DEFAULT,  get_global_default_encryption_key_id_value(), false); // TODO:Robert : zmiana na zero key_id - będzie to trzeba zmienić
+
+	if (crypt_data == NULL || crypt_data->key_found == false) {
+		mutex_exit(&fil_crypt_threads_mutex);
+		return false;
+	}
+
+	crypt_data->type = CRYPT_SCHEME_UNENCRYPTED;
+	crypt_data->min_key_version = ENCRYPTION_KEY_VERSION_NOT_ENCRYPTED; // all pages are unencrypted
+	crypt_data->rotate_state.start_time = time(0);
+	crypt_data->rotate_state.starting = true;
+	crypt_data->rotate_state.active_threads = 1;
+
+	if (space->encryption_type == Encryption::AES) {// We are re-encrypting space from MK encryption to RK encryption
+		crypt_data->encryption_rotation = Encryption::MASTER_KEY_TO_ROTATED_KEY;
+		ut_ad(space->encryption_key != NULL && space->encryption_iv != NULL);
+		crypt_data->set_tablespace_key(space->encryption_key);
+		crypt_data->set_tablespace_iv(space->encryption_iv); //space key and encryption are always initalized for MK encrypted tables
+	}
+
+	crypt_data->encrypting_with_key_version = crypt_data->key_get_latest_version();
+	ut_ad(crypt_data->encrypting_with_key_version != 0 && crypt_data->encrypting_with_key_version != ENCRYPTION_KEY_VERSION_INVALID);
+
+	if (crypt_data->key_found == false || crypt_data->load_needed_keys_into_local_cache() == false) {
+		// This should not happen, we have locked the keyring before encryption threads could have even started
+		// unless something realy strange have happend like removing keyring file from under running server.
+		ib::error() << "Encryption thread could not retrieve a key from a keyring for tablespace " << space->name
+			    << " . Removing space from encrypting. Please make sure keyring is functional and try restarting the server";
+		space->exclude_from_rotation = true;
+		mutex_exit(&fil_crypt_threads_mutex);
+		fil_space_destroy_crypt_data(&crypt_data);
+		return false;
+	}
+	mutex_enter(&crypt_data->mutex);
+	crypt_data = fil_space_set_crypt_data(space, crypt_data);
+	mutex_exit(&crypt_data->mutex);
+
+	space->encryption_type= Encryption::ROTATED_KEYS; // This works like this - if Encryption::ROTATED_KEYS is set - it means that
+	fil_crypt_start_converting = true;
+	mutex_exit(&fil_crypt_threads_mutex);
+
+	do
+	{
+		mtr_t mtr;
+		mtr.start();
+		mtr.set_named_space(space);
+
+		/* 2 - get page 0 */
+
+		buf_block_t* block = buf_page_get_gen(
+			page_id_t(space->id, 0), page_size_t(space->flags),
+			RW_X_LATCH, NULL, BUF_GET,
+			__FILE__, __LINE__,
+			&mtr);
+
+		/* 3 - write crypt data to page 0 */
+		byte* frame = buf_block_get_frame(block);
+		crypt_data->type = CRYPT_SCHEME_1;
+		crypt_data->write_page0(space, frame, &mtr, crypt_data->min_key_version, crypt_data->type, crypt_data->encryption_rotation);
+
+		mtr.commit();
+
+		/* 4 - sync tablespace before publishing crypt data */
+		buf_flush_request_force(LSN_MAX);
```

Comment:
I don't see space_id passed to this function. So it is not specific to tablespace. So are we flushing upto LSN_MAX.. Woah! This is very dangerous